src/mem/freelist.h

```diff
@@ -0,0 +1,297 @@
+#pragma once
+/**
+ * This file encapsulates the in disused object free lists
+ * that are used per slab.
+ */
+
+#include "../ds/address.h"
+#include "../ds/cdllist.h"
+#include "../ds/dllist.h"
+#include "../ds/helpers.h"
+#include "allocconfig.h"
+
+#include <iostream>
+
+namespace snmalloc
+{
+#ifdef CHECK_CLIENT
+  /**
+   * The key that is used to encode free list pointers.
+   * This should be randomised at startup in the future.
+   */
+  inline static uintptr_t global_key =
+    static_cast<size_t>(bits::is64() ? 0x9999'9999'9999'9999 : 0x9999'9999);
+
+  /**
+   * Used to turn a location into a key.  This is currently
+   * just the value of the previous location + 1.
+   */
+  inline static uintptr_t initial_key(void* p)
+  {
+    return address_cast(p) + 1;
+  }
+#endif
+
+  static inline bool different_slab(uintptr_t p1, uintptr_t p2)
+  {
+    return ((p1 ^ p2) >= SLAB_SIZE);
+  }
+
+  static inline bool different_slab(uintptr_t p1, void* p2)
+  {
+    return different_slab(p1, address_cast(p2));
+  }
+
+  static inline bool different_slab(void* p1, void* p2)
+  {
+    return different_slab(address_cast(p1), address_cast(p2));
+  }
+
+  /**
+   * Free objects within each slab point directly to the next.
+   * The next_object pointer can be encoded to detect
+   * corruption caused by writes in a UAF or a double free.
```

Comment:
Perhaps add commentary, assuming it's accurate, that `FreeObject`s are cache-friendly-offsetted within objects and so overlap `Remote`s?  Since we're eventually going to want (a configuration option) to zero on the `dealloc` side, it's important to note where we might deliberately put nonzeros and have to clear them on the `alloc` side.

Reply:
Added some comments